### PR TITLE
CBL-4008: [swift] FullTextIndexExpression

### PR DESCRIPTION
**Pull request by jayahariv**
* Implement the full text index expression protocol and related
* reference: f6ce1ba5e656418cfe837d671ee86c2e37d5f701 & #3085

### CODE REVIEW

File: Swift/FullTextIndexExpression.swift [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    func from(_ alias: String) -> FullTextIndexExpressionProtocol {
+        self.alias = alias
+        return self
+    }
+    
+    func stringVal() -> String {
+        if let alias = self.alias {
+            return "\(alias).\(self.name)"
+        }
+
```
Comment (jayahariv) [inline Swift/FullTextIndexExpression.swift:52]:
```suggestion
```

File: Swift/FullTextFunction.swift [hunk 2]
```diff
@@ -28,17 +28,47 @@ public final class FullTextFunction {
     ///
     /// - Parameter indexName: The index name.
     /// - Returns: The full-text rank function.
+    @available(*, deprecated, message: "Use FullTextFunction.rank(withIndex:) instead.")
     public static func rank(_ indexName: String) -> ExpressionProtocol {
-        return QueryExpression(CBLQueryFullTextFunction.rank(indexName))
+        return QueryExpression(CBLQueryFullTextFunction.rank(withIndexName: indexName))
```
Comment (pasin) [inline Swift/FullTextFunction.swift:33]: It's better to call to the new function?

File: Swift/FullTextFunction.swift [hunk 3]
```diff
@@ -28,17 +28,47 @@ public final class FullTextFunction {
...
     }
     
     /// Creates a full-text match expression with the given full-text index name and the query text
     ///
     /// - Parameter indexName: The index name.
     /// - Parameter query: The query string.
     /// - Returns: The full-text match function expression.
+    @available(*, deprecated, message: "Use FullTextFunction.match(withIndex:query) instead.")
     public static func match(indexName: String, query: String) -> ExpressionProtocol {
         return QueryExpression(CBLQueryFullTextFunction.match(withIndexName: indexName,
```
Comment (pasin) [inline Swift/FullTextFunction.swift:43]: Same as above, I think it's better to call the new function.

File: Swift/FullTextFunction.swift (hunk 2 above)
Comment (pasin) [reply Swift/FullTextFunction.swift:33]: The objective-c version should do the same as well.

File: Swift/Tests/DatabaseTest.swift [hunk 4]
```diff
@@ -1162,6 +1162,84 @@ class DatabaseTest: CBLTestCase {
...
+            .where(FullTextFunction.match(qualifiedIndex, query: "cat"))
+        
+        rows = try verifyQuery(query, block: { (n, r) in
+            XCTAssertEqual(r.string(at: 0), "doc\(n)")
+        })
+        XCTAssertEqual(rows, 2);
+    }
+    
+    // TODO: https://issues.couchbase.com/browse/CBL-3994
+    func testFTSQueryWithJoin() throws {
```
Comment (pasin) [inline Swift/Tests/DatabaseTest.swift:1206]: https://issues.couchbase.com/browse/CBL-3994 is fixed. Is TODO is still required?

File: Swift/FullTextFunction.swift (hunk 2 above)
Comment (jayahariv) [reply Swift/FullTextFunction.swift:33]: you mean get the Expression.fullTextIndex(name:), and pass it to the new API?

File: Swift/FullTextFunction.swift [hunk 5]
```diff
@@ -28,17 +28,55 @@ public final class FullTextFunction {
     ///
     /// - Parameter indexName: The index name.
     /// - Returns: The full-text rank function.
+    @available(*, deprecated, message: "Use FullTextFunction.rank(withIndex:) instead.")
     public static func rank(_ indexName: String) -> ExpressionProtocol {
-        return QueryExpression(CBLQueryFullTextFunction.rank(indexName))
+        let index = Expression.fullTextIndex(indexName)
+        guard let i = index as? FullTextIndexExpression else {
```
Comment (pasin) [inline Swift/FullTextFunction.swift:34]: Should be just `return rank(Expression.fullTextIndex(indexName))`?

File: Swift/FullTextFunction.swift [hunk 6]
```diff
@@ -28,17 +28,55 @@ public final class FullTextFunction {
...
     /// Creates a full-text match expression with the given full-text index name and the query text
     ///
     /// - Parameter indexName: The index name.
     /// - Parameter query: The query string.
     /// - Returns: The full-text match function expression.
+    @available(*, deprecated, message: "Use FullTextFunction.match(withIndex:query) instead.")
     public static func match(indexName: String, query: String) -> ExpressionProtocol {
-        return QueryExpression(CBLQueryFullTextFunction.match(withIndexName: indexName,
+        let index = Expression.fullTextIndex(indexName)
+        guard let i = index as? FullTextIndexExpression else {
```
Comment (pasin) [inline Swift/FullTextFunction.swift:48]: Should be just `return match(Expression.fullTextIndex(indexName), query: query)`?

**Review by pasin — CHANGES_REQUESTED** [review]

File: Swift/FullTextFunction.swift (hunk 6 above)
Comment (jayahariv) [reply Swift/FullTextFunction.swift]: cleaner!

**Review by pasin — APPROVED** [review]
Note: I have seen on the other PR for EE repo that add a header to CouchbaseLite.h. Is that also required here as well?

**Comment by jayahariv** [comment]
Yes, it required.

**Pull request merged**